devito/petsc/iet/routines.py

```diff
@@ -440,85 +698,128 @@ def runsolve(self, solver_objs, objs, rhs_callback, injectsolve):
-        and not (i.is_Dimension and not isinstance(i, (TimeDimension, ModuloDimension)))
-    ]
-    return petsc_struct(name, fields, liveness)
-
-
-def time_dep_replace(injectsolve, solver_objs, objs, sregistry):
-    target = injectsolve.expr.lhs
-    target_time = [
-        i for i, d in zip(target.indices, target.dimensions) if d.is_Time
-    ]
-    assert len(target_time) == 1
-    target_time = target_time.pop()
-
-    start_ptr = solver_objs['start_ptr']
-
-    vec_get_size = petsc_call(
-        'VecGetSize', [solver_objs['x_local'], Byref(solver_objs['localsize'])]
-    )
-
-    field_from_ptr = FieldFromPointer(
-        target.function._C_field_data, target.function._C_symbol
-    )
-
-    expr = DummyExpr(
-        start_ptr, cast_mapper[(target.dtype, '*')](field_from_ptr) +
-        Mul(target_time, solver_objs['localsize']), init=True
-    )
-
-    vec_replace_array = petsc_call('VecReplaceArray', [solver_objs['x_local'], start_ptr])
-    return (vec_get_size, expr, vec_replace_array)
-
-
-def uxreplace_time(body, solver_objs):
-    # TODO: Potentially introduce a TimeIteration abstraction to simplify
-    # all the time processing that is done (searches, replacements, ...)
-    # "manually" via free functions
-    time_spacing = solver_objs['target'].grid.stepping_dim.spacing
-    true_dims = solver_objs['true_dims']
-
-    time_mapper = {
-        v: k.xreplace({time_spacing: 1, -time_spacing: -1})
-        for k, v in solver_objs['time_mapper'].items()
-    }
-    subs = {symb: true_dims[time_mapper[symb]] for symb in time_mapper}
-    return Uxreplace(subs).visit(body)
+            field_from_ptr = FieldFromPointer(
+                target.function._C_field_data, target.function._C_symbol
+            )
+
+            expr = DummyExpr(
+                start_ptr, cast_mapper[(target.dtype, '*')](field_from_ptr) +
+                Mul(target_time, solver_objs['localsize']), init=True
+            )
+
+            vec_replace_array = petsc_call(
+                'VecReplaceArray', [solver_objs['x_local'], start_ptr]
+            )
+            return (vec_get_size, expr, vec_replace_array)
+        else:
+            field_from_ptr = FieldFromPointer(
+                target.function._C_field_data, target.function._C_symbol
+            )
+            vec_replace_array = (petsc_call(
+                'VecReplaceArray', [solver_objs['x_local'], field_from_ptr]
+            ),)
+            return vec_replace_array
+
+    def assign_time_iters(self, iters, struct):
+        """
+        Assign time iterators to the struct.
+        Ensure that assignment occurs only once per time loop, if necessary.
+        Assign only the iterators that are common between the struct fields
+        and the actual Iteration.
+        """
+        time_iter = [
+            i for i in FindNodes(Iteration).visit(iters)
+            if i.dim.is_Time
+        ]
+        assert len(time_iter) == 1
+        time_iter, = time_iter
```

Comment:
Nitpick: `time_iter = time_iter.pop()` is marginally more readable and more consistent with other functions